portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -27,51 +34,92 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  // default values without user input
-  private ArrayList<String> messages = new ArrayList<String>(
-      Arrays.asList("No comments")
-  );
+  private ArrayList<Comment> messages = new ArrayList<>();
+  private String homePage = "/index.html";
+  //properties of the Comments
+  private String classType = "Comment";
+  private String commentTimestamp = "timestampUTC";
+  private String commentName = "name";
+  private String commentEmail = "email";
+  private String commentSubject = "subject";
+  private String fullComments = "comments";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    
-    // Convert the arraylist to JSON
-    String json = convertToJson(messages);
 
-    // Send the JSON as the response
+    Query query = new Query(classType).addSort(commentTimestamp, SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    String numResultsString = (request.getQueryString());
+    int numResults;
+    if (numResultsString == null) {
+      numResults = 1;
+    } else if (numResultsString.equals("num-results=1")) {
+      numResults = 1;
+    } else if (numResultsString.equals("num-results=2")) {
+      numResults = 2;
+    } else if (numResultsString.equals("num-results=3")) {
+      numResults = 3;
+    } else if (numResultsString.equals("num-results=4")) {
+      numResults = 4;
+    } else {
+      numResults = 1;
+    }
+
+    ArrayList<Comment> commentList = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
```

Comment:
To select n elements from an Iterable, you could use guava utility method (you can do this in future PR).

https://stackoverflow.com/questions/38128230/get-first-n-elements-of-an-iterable
https://guava.dev/releases/19.0/api/docs/com/google/common/collect/Iterables.html#limit(java.lang.Iterable,%20int)